File: sdk/rust/tests/test_result.rs
//! Tests for the0_sdk::input::success() and result() functions
//!
//! These functions write to a result file, not stdout.

mod common;

use common::ResultContext;
use serde_json::json;
use the0_sdk::input;

#[test]
fn success_writes_json_to_result_file() {
    let ctx = ResultContext::new();
    input::success("Test message");

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert_eq!(parsed["status"], "success");
    assert_eq!(parsed["message"], "Test message");
}

#[test]
fn success_escapes_quotes() {
    let ctx = ResultContext::new();
    input::success(r#"Message with "quotes""#);

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert_eq!(parsed["message"], r#"Message with "quotes""#);
}

#[test]
fn success_escapes_backslashes() {
    let ctx = ResultContext::new();
    input::success(r#"Path: C:\Users\test"#);

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert_eq!(parsed["message"], r#"Path: C:\Users\test"#);
}

#[test]
fn success_handles_empty_message() {
    let ctx = ResultContext::new();
    input::success("");

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert_eq!(parsed["status"], "success");
    assert_eq!(parsed["message"], "");
}

#[test]
fn success_handles_long_message() {
    let ctx = ResultContext::new();
    let long_message = "x".repeat(10000);
    input::success(&long_message);

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert_eq!(parsed["message"].as_str().unwrap().len(), 10000);
}

#[test]
fn result_writes_custom_json() {
    let ctx = ResultContext::new();
    let data = json!({
        "status": "success",
        "trade_id": "abc123",
        "amount": 100.5
    });
    input::result(&data);

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert_eq!(parsed["status"], "success");
    assert_eq!(parsed["trade_id"], "abc123");
    assert_eq!(parsed["amount"], 100.5);
}

#[test]
fn result_writes_nested_objects() {
    let ctx = ResultContext::new();
    let data = json!({
        "status": "success",
        "data": {
            "price": 45000.0,
            "volume": 1.5
        }
    });
    input::result(&data);

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert_eq!(parsed["data"]["price"], 45000.0);
    assert_eq!(parsed["data"]["volume"], 1.5);
}

#[test]
fn result_writes_arrays() {
    let ctx = ResultContext::new();
    let data = json!({
        "trades": ["trade1", "trade2", "trade3"]
    });
    input::result(&data);

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert!(parsed["trades"].is_array());
    assert_eq!(parsed["trades"].as_array().unwrap().len(), 3);
}

#[test]
fn result_writes_empty_object() {
    let ctx = ResultContext::new();
    input::result(&json!({}));

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert!(parsed.is_object());
    assert!(parsed.as_object().unwrap().is_empty());
}

#[test]
fn result_overwrites_previous_result() {
    let ctx = ResultContext::new();

    input::success("First message");
    let content1 = ctx.read_result().expect("Result file should exist");
    let parsed1: serde_json::Value = serde_json::from_str(&content1).unwrap();
    assert_eq!(parsed1["message"], "First message");

    input::success("Second message");
    let content2 = ctx.read_result().expect("Result file should exist");
    let parsed2: serde_json::Value = serde_json::from_str(&content2).unwrap();
    assert_eq!(parsed2["message"], "Second message");
}

#[test]
fn success_creates_valid_json() {
    let ctx = ResultContext::new();
    input::success("Test");

    let content = ctx.read_result().expect("Result file should exist");

    // Should not panic
    let _: serde_json::Value = serde_json::from_str(&content)
        .expect("Result should be valid JSON");
}

#[test]
fn result_handles_special_characters() {
    let ctx = ResultContext::new();
    let data = json!({
        "message": "Line1\nLine2\tTabbed",
        "path": "C:\\Users\\test"
    });
    input::result(&data);

    let content = ctx.read_result().expect("Result file should exist");
    let parsed: serde_json::Value = serde_json::from_str(&content).unwrap();

    assert_eq!(parsed["message"], "Line1\nLine2\tTabbed");
    assert_eq!(parsed["path"], "C:\\Users\\test");
}

File: sdk/cpp/tests/test_result.cpp
/**
 * Tests for the0::success(), the0::result() functions
 *
 * These functions write results to a file (result.json), not stdout.
 * Tests use ResultFileContext to set up a temp directory.
 */

#include <doctest/doctest.h>
#include <the0/the0.h>
#include "helpers/test_helpers.hpp"

using namespace the0_test;

TEST_SUITE("the0::success") {
    TEST_CASE("writes valid JSON with status success to result file") {
        ResultFileContext ctx;
        the0::success("Test message");

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["status"] == "success");
        CHECK(json["message"] == "Test message");
    }

    TEST_CASE("properly escapes special characters") {
        ResultFileContext ctx;
        the0::success("Message with \"quotes\" and \\backslash");

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["message"] == "Message with \"quotes\" and \\backslash");
    }

    TEST_CASE("handles newlines and tabs") {
        ResultFileContext ctx;
        the0::success("Line1\nLine2\tTabbed");

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["message"] == "Line1\nLine2\tTabbed");
    }

    TEST_CASE("handles unicode characters") {
        ResultFileContext ctx;
        the0::success("Hello World!");

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["message"] == "Hello World!");
    }

    TEST_CASE("handles empty message") {
        ResultFileContext ctx;
        the0::success("");

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["status"] == "success");
        CHECK(json["message"] == "");
    }

    TEST_CASE("handles very long messages") {
        ResultFileContext ctx;
        std::string long_message(10000, 'x');
        the0::success(long_message);

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["message"].get<std::string>().size() == 10000);
    }
}

TEST_SUITE("the0::result with json") {
    TEST_CASE("writes custom JSON object to result file") {
        ResultFileContext ctx;
        the0::result({
            {"status", "success"},
            {"trade_id", "abc123"},
            {"amount", 100.5}
        });

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["status"] == "success");
        CHECK(json["trade_id"] == "abc123");
        CHECK(json["amount"] == 100.5);
    }

    TEST_CASE("handles nested objects") {
        ResultFileContext ctx;
        the0::result({
            {"status", "success"},
            {"data", {
                {"price", 45000.0},
                {"volume", 1.5}
            }}
        });

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["data"]["price"] == 45000.0);
        CHECK(json["data"]["volume"] == 1.5);
    }

    TEST_CASE("handles arrays") {
        ResultFileContext ctx;
        the0::result({
            {"trades", {"trade1", "trade2", "trade3"}}
        });

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["trades"].is_array());
        CHECK(json["trades"].size() == 3);
    }

    TEST_CASE("handles empty object") {
        ResultFileContext ctx;
        the0::result(the0::json::object());

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json.is_object());
        CHECK(json.empty());
    }
}

TEST_SUITE("the0::result with status/message/data") {
    TEST_CASE("creates result with status and message") {
        ResultFileContext ctx;
        the0::result("success", "Trade completed");

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["status"] == "success");
        CHECK(json["message"] == "Trade completed");
    }

    TEST_CASE("merges additional data fields") {
        ResultFileContext ctx;
        the0::result("success", "Trade executed", {
            {"trade_id", "12345"},
            {"price", 45000.50},
            {"quantity", 0.1}
        });

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["status"] == "success");
        CHECK(json["message"] == "Trade executed");
        CHECK(json["trade_id"] == "12345");
        CHECK(json["price"] == 45000.50);
        CHECK(json["quantity"] == 0.1);
    }

    TEST_CASE("works with empty data") {
        ResultFileContext ctx;
        the0::result("error", "Something failed", {});

        std::string output = ctx.read_result();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["status"] == "error");
        CHECK(json["message"] == "Something failed");
        CHECK(json.size() == 2); // Only status and message
    }
}
